raiden/raiden_service.py

```diff
@@ -540,9 +541,67 @@ def connection_manager_for_token(self, token_address):
             raise InvalidAddress('token is not registered.')
         return manager
 
+    def leave_all_token_networks_async(self):
+        token_addresses = self.channelgraphs.keys()
+        leave_results = []
+        for token_address in token_addresses:
+            try:
+                connection_manager = self.connection_manager_for_token(token_address)
+            except InvalidAddress:
+                pass
+            leave_results.append(connection_manager.leave_async())
+        return leave_results
+
+    def close_and_settle(self):
+        log.info('raiden will close and settle all channels now')
+
+        connection_managers = [
+            self.connection_manager_for_token(token_address) for
+            token_address in self.channelgraphs.keys()
+        ]
+
+        def blocks_to_wait():
+            return max(
+                connection_manager.min_settle_blocks
+                for connection_manager in connection_managers
+            )
+
+        all_channels = sum(
+            [connection_manager.open_channels for connection_manager in connection_managers], []
+        )
+
+        leave_greenlets = self.leave_all_token_networks_async()
+        # using the un-cached block number here
+        last_block = self.chain.block_number()
+
+        earliest_settlement = last_block + blocks_to_wait()
+
+        current_block = last_block
+        while current_block < earliest_settlement:
+            gevent.sleep(self.alarm.wait_time)
+            current_block = self.chain.block_number()
+            log.info(
+                'waiting at least %s more blocks for settlement (%s channels not yet settled)' % (
+                    blocks_to_wait(),
+                    sum(
```

Comment:
I think it would just make the code a bit less verbose but it's up to you. It's your PR and this comment is of the "opinion" type so it's all up to you.